runtime/src/bank.rs

```diff
@@ -7316,14 +7332,49 @@ impl Bank {
         new_feature_activations: &HashSet<Pubkey>,
     ) {
         for builtin in BUILTINS.iter() {
-            if let Some(feature_id) = builtin.feature_id {
-                let should_apply_action_for_feature_transition =
-                    if only_apply_transitions_for_new_features {
+            // The `builtin_disabled` flag is used to handle the case where a
+            // built-in is scheduled to be enabled by one feature gate and
+            // later migrated to Core BPF by another.
+            // There should never be a case where a built-in is set to be
+            // migrated to Core BPF and is also set to be enabled on feature
+            // activation on the same feature gate. However, the
+            // `builtin_disabled` flag will handle this case as well, electing
+            // to first attempt the migration to Core BPF.
+            // This will fail gracefully, and the built-in will subsequently be
+            // enabled, but it will never be migrated to Core BPF.
+            // Using the same feature gate for both enabling and migrating a
+            // built-in to Core BPF should be strictly avoided.
+            let mut builtin_disabled = false;
+            if let Some(core_bpf_migration) = &builtin.core_bpf_migration {
+                // If the built-in is set to be migrated to Core BPF on feature
+                // activation, perform the migration and do not add the program
+                // to the bank's builtins. The migration will remove it from
+                // the builtins list and the cache.
+                if new_feature_activations.contains(&core_bpf_migration.feature_id) {
+                    if let Err(e) = builtin.migrate_to_core_bpf(self) {
+                        warn!(
+                            "Failed to migrate built-in {} to Core BPF: {}",
+                            builtin.name, e
+                        );
+                    } else {
+                        builtin_disabled = true;
+                    }
+                } else {
+                    // If the built-in has already been migrated to Core BPF, do not
+                    // add it to the bank's builtins.
+                    builtin_disabled = self.feature_set.is_active(&core_bpf_migration.feature_id);
```

Comment:
Yep! The idea was that we'd do that on feature cleanup.

Reply:
I'd like some extra eyes on this bit from the runtime team.

For typical Banking Stage and Replay Stage, removing a program from the static `BUILTINS` list won't cause a problem, thanks to the existing `BankForks`.

The programs cache and the list of builtin program IDs are passed from the parent bank to the new bank directly via `_new_from_parent` [right here](https://github.com/anza-xyz/agave/blob/64412096829241e5c70d537403584dbc9e04ee93/runtime/src/bank.rs#L1314) and [right here](https://github.com/anza-xyz/agave/blob/64412096829241e5c70d537403584dbc9e04ee93/runtime/src/bank.rs#L1229-L1230). The cache is then recompiled, checking if any program that needs recompilation is a builtin, which will no longer be the case for our target since the owner (loader) has changed and we've popped it from the cache on migration [here](https://github.com/buffalojoec/solana/blob/79574392409d61c4990d5e6deb70d35ead175bc0/runtime/src/bank/builtins/core_bpf_migration/mod.rs#L113-L117).

However, reading a new `BankForks` instance from blockstore may be problematic. Creating a bank for slot 0, ie. [`ledger::blockstore_processor::process_blockstore_for_bank_0`](https://github.com/anza-xyz/agave/blob/64412096829241e5c70d537403584dbc9e04ee93/ledger/src/blockstore_processor.rs#L788) will use `Bank::new_with_paths(..)`. If the builtin is missing when Bank0 is attempting to be created, that would result in a different `BankForks` and subsequently a different block replay result if replayed from genesis.

Perhaps removing the list altogether - which we've been discussing - could remedy this issue. Investigating.